### ISSUE TITLE
Question about the privacy cost calculation

### ISSUE THREAD

**Issue by lsl001006** [DingfanChen/GS-WGAN#11]
Hi Dingfan! Sorry to bother you again.

When I'm evaluating the privacy cost of the GS-WGAN code from [GS-WGAN/evaluation/privacy_analysis.py](https://github.com/DingfanChen/GS-WGAN/blob/main/evaluation/privacy_analysis.py):
```
def main(config):
    delta = 1e-5
    batch_size = config['batchsize']
    prob = 1. / config['num_discriminators']  # subsampling rate
    n_steps = config['iterations']  # training iterations
    sigma = config['noise_multiplier']  # noise scale
    func = lambda x: rdp_bank.RDP_gaussian({'sigma': sigma}, x)

    acct = rdp_acct.anaRDPacct()
    acct.compose_subsampled_mechanism(func, prob, coeff=n_steps * batch_size)
    epsilon = acct.get_eps(delta)
    print("Privacy cost is: epsilon={}, delta={}".format(epsilon, delta))
```
It seems that the final privacy cost is about the calculation of parameter `delta, batch_size, num_discriminators, iterations, and noise_multiplier`, but in the `dp_conv_hook` module in [main.py](https://github.com/DingfanChen/GS-WGAN/blob/main/source/main.py):
```
def dp_conv_hook(module, grad_input, grad_output):
    '''
    gradient modification + noise hook
    :param module:
    :param grad_input:
    :param grad_output:
    :return:
    '''
    global noise_multiplier
    ### get grad wrt. input (image)
    grad_wrt_image = grad_input[0]
    grad_input_shape = grad_wrt_image.size()
    batchsize = grad_input_shape[0]
    clip_bound_ = CLIP_BOUND / batchsize

    grad_wrt_image = grad_wrt_image.view(batchsize, -1)
    grad_input_norm = torch.norm(grad_wrt_image, p=2, dim=1)

    ### clip
    clip_coef = clip_bound_ / (grad_input_norm + 1e-10)
    clip_coef = torch.min(clip_coef, torch.ones_like(clip_coef))
    clip_coef = clip_coef.unsqueeze(-1)
    grad_wrt_image = clip_coef * grad_wrt_image

    ### add noise
    noise = clip_bound_ * noise_multiplier * SENSITIVITY * torch.randn_like(grad_wrt_image)
    grad_wrt_image = grad_wrt_image + noise
    grad_input_new = [grad_wrt_image.view(grad_input_shape)]
    for i in range(len(grad_input) - 1):
        grad_input_new.append(grad_input[i + 1])
    return tuple(grad_input_new)
```
It seems that there are also hyperparameters like `CLIP_BOUND`, and `SENSITIVITY` associated with the backward loss, so I'm confused about 2 questions as follows:
1. when calculating privacy cost, why are the hyperparams like  `CLIP_BOUND`,`SENSITIVITY` not included in the privacy_analysis function?
2. why should we define the variable `prob` in privacy_analysis.py's main function as `1/config['num_discriminators']` ? Does this mean that privacy problems may occur when `num_discriminators` is small?

Thank you very much for reading the above, and I'm looking forward to your reply if you got time to answer this.

**Comment by DingfanChen** [comment]
Hi, thanks for your questions. 

1. Briefly speaking, ```CLIP_BOUND``` determines ```SENSITIVITY```, and ```SENSITIVITY``` determines the scale of DP noise (noise scale $\propto$ ```SENSITIVITY``` $\times$  ```noise_multiplier```). Among all the hyperparameters mentioned above, only the noise-to-signal ratio (i.e., ```noise_multiplier```) will affect the privacy level. For this point, I would recommend the [DP-SGD paper](https://arxiv.org/pdf/1607.00133.pdf) for a more in-depth explanation. 

2. ```1/config['num_discriminators']``` corresponds to the sub-sampling rate. As we also mentioned in the paper (Figure 4 (a)), the num_discriminators should be sufficiently large (>500) in order to have a good privacy level.
